editor/store/selectors.js

```diff
@@ -1196,152 +1203,280 @@ export function getNotices( state ) {
-	if ( blockTypeIsDisabled ) {
-		return null;
-	}
+		const parentBlockListSettings = getBlockListSettings( state, parentUID );
+		const parentAllowedBlocks = get( parentBlockListSettings, [ 'supportedBlocks' ] );
+		const hasParentAllowedBlock = checkAllowList( parentAllowedBlocks, blockName );
+
+		const blockAllowedParentBlocks = blockType.parent;
+		const parentName = getBlockName( state, parentUID );
+		const hasBlockAllowedParent = checkAllowList( blockAllowedParentBlocks, parentName );
+
+		let isBlockAllowedInParent;
+		if ( hasParentAllowedBlock !== null && hasBlockAllowedParent !== null ) {
+			isBlockAllowedInParent = hasParentAllowedBlock || hasBlockAllowedParent;
+		} else if ( hasParentAllowedBlock !== null ) {
+			isBlockAllowedInParent = hasParentAllowedBlock;
+		} else if ( hasBlockAllowedParent !== null ) {
+			isBlockAllowedInParent = hasBlockAllowedParent;
+		} else {
+			isBlockAllowedInParent = true;
+		}
 
-	const referencedBlock = getBlock( state, sharedBlock.uid );
-	if ( ! referencedBlock ) {
-		return null;
-	}
+		if ( ! isBlockAllowedInParent ) {
+			return false;
+		}
 
-	const referencedBlockType = getBlockType( referencedBlock.name );
-	if ( ! referencedBlockType ) {
-		return null;
-	}
+		return true;
+	},
+	( state, blockName, parentUID ) => [
+		state.blockListSettings[ parentUID ],
+		state.editor.present.blocksByUID[ parentUID ],
+		state.settings.allowedBlockTypes,
+		state.settings.templateLock,
+	],
+);
 
-	return {
-		id: `core/block/${ sharedBlock.id }`,
-		name: 'core/block',
-		initialAttributes: { ref: sharedBlock.id },
-		title: sharedBlock.title,
-		icon: referencedBlockType.icon,
-		category: 'shared',
-		keywords: [],
-		isDisabled: false,
-	};
+/**
+ * Returns information about how recently and frequently a block has been inserted.
+ *
+ * @param {Object} state Global application state.
+ * @param {string} id    A string which identifies the insert, e.g. 'core/block/12'
+ *
+ * @return {?{ time: number, count: number }} An object containing `time` which is when the last
+ *                                            insert occured as a UNIX epoch, and `count` which is
+ *                                            the number of inserts that have occurred.
+ */
+function getInsertUsage( state, id ) {
+	return state.preferences.insertUsage[ id ] || null;
 }
 
 /**
  * Determines the items that appear in the the inserter. Includes both static
  * items (e.g. a regular block type) and dynamic items (e.g. a shared block).
  *
- * @param {Object}           state             Global application state.
- * @param {string[]|boolean} allowedBlockTypes Allowed block types, or true/false to enable/disable all types.
+ * Each item object contains what's necessary to display a button in the
+ * inserter and handle its selection.
+ *
+ * The 'utility' property indicates how useful we think an item will be to the
```

Comment:
Can you explain the difference between `utility` as defined here and `priority` as defined by [`@wordpress/hooks`](https://npmjs.com/package/@wordpress/hooks) and the [block type `transforms` property](https://wordpress.org/gutenberg/handbook/block-api/#transforms-optional)?

Reply:
>Not sure that I understand your question, sorry. Would you prefer that I rename `utility` → `priority`? I named it `utility` because it's a measure of _how **useful** an inserter item is_, but I'm not tied to the name.

I'm curious more in hopes of minimizing the number of concepts which exist, where an existing pattern already exists to serve a need. As I read further, I see that these are used as specific multipliers, so while similar to priority, not quite the same.

Which is to say: Nothing should need to be changed here.

Reply:
At some point we might want to consolidate those two properties: utility and frecency into one property called score. This is how some search engines work. The score could be calculated using [weighted sum model](https://en.m.wikipedia.org/wiki/Weighted_sum_model) to translate how things are sorted at the moment.